OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,144 @@
+
+		struct Text
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			public Color Color;
+			public string Team;		// Used only for team label
+			public string Player;
+			public string Power;
+			public string RemainingTime;
 		}
 
-		public override void Tick()
+		readonly List<Text> display;
+		readonly bool noPowers;
+		bool alternateColor, needTeamLabel, showPlayer;
+
+		readonly World world;
+		int lastTick;
+
+		[ObjectCreator.UseCtor]
+		public SupportPowerTimerWidget(World world)
 		{
-			texts = powers.Select(p =>
-			{
-				var time = WidgetUtils.FormatTime(p.RemainingTime, false);
-				var text = Format.F(p.Info.Description, time);
-				var color = !p.Ready || Game.LocalTick % 50 < 25 ? p.Instances[0].Self.Owner.Color.RGB : Color.White;
-				return Pair.New(text, color);
-			}).ToArray();
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			var availablePowers = world.Map.Rules.Actors.Values
+				.SelectMany(ai => ai.Traits.WithInterface<SupportPowerInfo>())
+				.Where(i => i.DisplayTimer).ToArray();
+
+			noPowers = !spManagers.Keys.Any() || !availablePowers.Any();
+			if (noPowers)
+				return;
+
+			playersByTeam = new SortedDictionary<int, Player[]>();
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+				playersByTeam[team.Key] = team.ToArray();
+
+			font = Game.Renderer.Fonts[fontName];
+			yIncrement = font.Measure(" ").Y + 5;
+			var padding = font.Measure("  ").X;
+			var maxPlayerWidth = spManagers.Keys.Max(p => font.Measure(p.PlayerName).X);
+			var maxPowerWidth = availablePowers.Max(i => font.Measure(i.Description).X);
+			powerPos = maxPlayerWidth + padding;
+			timePos = powerPos + maxPowerWidth + padding;
+
+			display = new List<Text>();
+			this.world = world;
 		}
 
 		public override void Draw()
 		{
-			if (!IsVisible() || texts == null)
+			if (!IsVisible() || noPowers)
 				return;
 
-			var y = 0;
-			foreach (var t in texts)
+			if (world.WorldTick - lastTick > 9)
 			{
-				var font = Game.Renderer.Fonts[Font];
-				font.DrawTextWithContrast(t.First, new float2(Bounds.Location) + new float2(0, y), t.Second, Color.Black, 1);
-				y += (font.Measure(t.First).Y + 5) * (int)Order;
+				lastTick = world.WorldTick;
+				display.Clear();
+				alternateColor = Game.LocalTick % 50 > 25;
+				foreach (var team in playersByTeam)
+				{
+					needTeamLabel = true;
+					foreach (var player in team.Value)
+					{
+						var powers = spManagers[player].Powers.Values
+							.Where(i => i.Info.DisplayTimer && i.Instances.Any() && !i.Disabled).ToArray();
```

Comment:
no. the support power will be destroyed before it is removed from the list.

On 2015-07-15, at 10:35 PM, Deniz <notifications@github.com> wrote:

> In OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
> 
> >  			{
> > -				var font = Game.Renderer.Fonts[Font];
> > -				font.DrawTextWithContrast(t.First, new float2(Bounds.Location) + new float2(0, y), t.Second, Color.Black, 1);
> > -				y += (font.Measure(t.First).Y + 5) * (int)Order;
> > +				lastTick = world.WorldTick;
> > +				display.Clear();
> > +				alternateColor = Game.LocalTick % 50 > 25;
> > +				foreach (var team in playersByTeam)
> > +				{
> > +					needTeamLabel = true;
> > +					foreach (var player in team.Value)
> > +					{
> > +						var powers = spManagers[player].Powers.Values
> > +							.Where(i => i.Info.DisplayTimer && i.Instances.Any() && !i.Disabled).ToArray();
> Does this change while it is still being queried and cause the NRE?
> 
> ―
> Reply to this email directly or view it on GitHub.
>